### PR TITLE
feat(api): manage dependencies during module install/update/remove

### PR DESCRIPTION
**Pull request by kduret**
## Description

* manage dependencies during module install/update/remove
* remove dead code

image

**Fixes** MON-12229

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 21.04.x
- [ ] 21.10.x
- [ ] 22.04.x
- [x] 22.10.x (master)

### CODE REVIEW

File: src/CentreonModule/Infrastructure/Entity/Module.php [hunk 1]
```diff
@@ -120,6 +120,11 @@ class Module implements SourceDataInterface
      */
     private $isUpdated = false;
 
+    /**
+     * @var string[]
```
Comment (kduret) [inline src/CentreonModule/Infrastructure/Entity/Module.php]: explain property

File: src/CentreonModule/Infrastructure/Entity/Module.php [hunk 2]
```diff
@@ -377,4 +382,20 @@ public function setUpdated(bool $value): void
...
+    {
+        return $this->dependencies;
+    }
+
+    /**
+     * @param string[] $dependencies
+     */
+    public function setDependencies(array $dependencies): void
+    {
+        $this->dependencies = $dependencies;
```
Comment (kduret) [inline src/CentreonModule/Infrastructure/Entity/Module.php]: validate data

Comment (kduret) [reply src/CentreonModule/Infrastructure/Entity/Module.php]: setdependencies should call addDependency

File: src/CentreonModule/Infrastructure/Source/ModuleSource.php [hunk 3]
```diff
@@ -262,4 +274,91 @@ protected function processLicense(string $moduleId, array $info): array
...
+
+        return $dependencies;
+    }
+
+    /**
+     * Validate requirements before remove (dependencies)
+     *
+     * @param string $moduleId (example: centreon-license-manager)
+     *
+     * @throws \Exception
```
Comment (kduret) [inline src/CentreonModule/Infrastructure/Source/ModuleSource.php]:
```suggestion
     * @throws ModuleException
```

File: src/CentreonModule/Infrastructure/Source/SourceAbstract.php [hunk 4]
```diff
@@ -111,6 +108,12 @@ public function update(string $id): ?Module
         return $this->getDetail($id);
     }
 
+    /**
+     * Remove module or widget
+     *
+     * @param string $id
+     * @return void
```
Comment (kduret) [inline src/CentreonModule/Infrastructure/Source/SourceAbstract.php]:
```suggestion
```

File: src/CentreonModule/Infrastructure/Source/ModuleSource.php [hunk 5]
```diff
@@ -262,4 +274,91 @@ protected function processLicense(string $moduleId, array $info): array
...
+    /**
+     * Sort module dependencies
+     *
+     * @param string $moduleId (example: centreon-license-manager)
+     * @param string[] $alreadyProcessed
+     * @return string[]
+     *
+     * @throws \Exception
+     */
+    private function getSortedDependencies(
```
Comment (kduret) [inline src/CentreonModule/Infrastructure/Source/ModuleSource.php:313]: add unit test

Comment (kduret) [reply src/CentreonModule/Infrastructure/Source/ModuleSource.php:313]: Too hard... cause of injected container

**Review by a-launois — APPROVED** [review]

**Review by TamazC — APPROVED** [review]

**Review by callapa — APPROVED** [review]

**Review by smau-centreon — APPROVED** [review]

**Pull request merged**